File: supabase/functions/chatbot-whatsapp-webhook/chatbot/profile_management.ts
import { ConversationState, updateConversationState } from './conversation_state.ts'
import { getCustomerProfile, updateCustomerProfile } from './database_helpers.ts'
import { 
  formatProfileView, 
  formatProfileEdit, 
  formatFieldEdit, 
  formatProfileUpdated,
  formatMainMenu,
  formatError 
} from './message_formatter.ts'

export async function handleProfileManagement(
  chatId: string,
  customerId: string,
  messageBody: string,
  state: ConversationState
): Promise<string> {
  switch (state.step) {
    case 'profile_view':
      return await handleProfileView(chatId, customerId, messageBody, state)
      
    case 'profile_edit':
      return await handleProfileEdit(chatId, customerId, messageBody, state)
      
    case 'profile_field_edit':
      return await handleProfileFieldEdit(chatId, customerId, messageBody, state)
      
    default:
      // Default to profile view
      return await handleProfileView(chatId, customerId, messageBody, state)
  }
}

async function handleProfileView(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  if (!isNaN(selectedNumber)) {
    switch (selectedNumber) {
      case 1: // Edit Profile
        const customer = await getCustomerProfile(customerId)
        if (!customer) {
          return formatError('Unable to load profile. Please try again.')
        }
        
        await updateConversationState(chatId, {
          ...state,
          step: 'profile_edit',
          previous_step: 'profile_view'
        })
        
        return formatProfileEdit(customer)
        
      case 2: // Back to Main Menu
        await updateConversationState(chatId, {
          ...state,
          step: 'main_menu',
          previous_step: 'profile_view'
        })
        
        return formatMainMenu()
        
      default:
        const customerForError = await getCustomerProfile(customerId)
        if (!customerForError) {
          return formatError('Unable to load profile. Please try again.')
        }
        
        const errorMessage = formatError('Please select 1 to edit profile or 2 for main menu')
        const profileView = formatProfileView(customerForError)
        return errorMessage + '\n\n' + profileView
    }
  }
  
  // Default: show profile view
  const customer = await getCustomerProfile(customerId)
  if (!customer) {
    return formatError('Unable to load profile. Please try again.')
  }
  
  return formatProfileView(customer)
}

async function handleProfileEdit(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  if (!isNaN(selectedNumber)) {
    const customer = await getCustomerProfile(customerId)
    if (!customer) {
      return formatError('Unable to load profile. Please try again.')
    }
    
    switch (selectedNumber) {
      case 1: // Edit First Name
        await updateConversationState(chatId, {
          ...state,
          step: 'profile_field_edit',
          editing_field: 'first_name',
          previous_step: 'profile_edit'
        })
        
        return formatFieldEdit('First Name', customer.first_name)
        
      case 2: // Edit Last Name
        await updateConversationState(chatId, {
          ...state,
          step: 'profile_field_edit',
          editing_field: 'last_name',
          previous_step: 'profile_edit'
        })
        
        return formatFieldEdit('Last Name', customer.last_name)
        
      case 3: // Edit Email
        await updateConversationState(chatId, {
          ...state,
          step: 'profile_field_edit',
          editing_field: 'email',
          previous_step: 'profile_edit'
        })
        
        return formatFieldEdit('Email Address', customer.email)
        
      case 4: // Back to Profile
        await updateConversationState(chatId, {
          ...state,
          step: 'profile_view',
          previous_step: 'profile_edit'
        })
        
        return formatProfileView(customer)
        
      case 5: // Back to Main Menu
        await updateConversationState(chatId, {
          ...state,
          step: 'main_menu',
          previous_step: 'profile_edit'
        })
        
        return formatMainMenu()
        
      default:
        const errorMessage = formatError('Please select a valid option (1-5)')
        const editView = formatProfileEdit(customer)
        return errorMessage + '\n\n' + editView
    }
  }
  
  // If not a number, show error with menu
  const customer = await getCustomerProfile(customerId)
  if (!customer) {
    return formatError('Unable to load profile. Please try again.')
  }
  
  const errorMessage = formatError('Please select a valid option')
  const editView = formatProfileEdit(customer)
  return errorMessage + '\n\n' + editView
}

async function handleProfileFieldEdit(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const input = messageBody.trim()
  
  // Check for cancel
  if (input.toLowerCase() === 'cancel') {
    const customer = await getCustomerProfile(customerId)
    if (!customer) {
      return formatError('Unable to load profile. Please try again.')
    }
    
    await updateConversationState(chatId, {
      ...state,
      step: 'profile_edit',
      editing_field: undefined,
      previous_step: 'profile_field_edit'
    })
    
    return formatProfileEdit(customer)
  }
  
  // Validate input based on field type
  if (state.editing_field === 'email' && input) {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
    if (!emailRegex.test(input)) {
      const customer = await getCustomerProfile(customerId)
      const errorMessage = formatError('Please enter a valid email address (e.g., yourname@email.com)')
      const fieldEdit = formatFieldEdit('Email Address', customer?.email)
      return errorMessage + '\n\n' + fieldEdit
    }
  }
  
  if (!input || input.length < 1) {
    const customer = await getCustomerProfile(customerId)
    const fieldName = state.editing_field === 'first_name' ? 'First Name' : 
                     state.editing_field === 'last_name' ? 'Last Name' : 'Email Address'
    const currentValue = state.editing_field === 'first_name' ? customer?.first_name :
                        state.editing_field === 'last_name' ? customer?.last_name : customer?.email
    
    const errorMessage = formatError('Please enter a value or type "cancel" to go back')
    const fieldEdit = formatFieldEdit(fieldName, currentValue)
    return errorMessage + '\n\n' + fieldEdit
  }
  
  // Update the profile
  const updates: any = {}
  let fieldDisplayName = ''
  
  switch (state.editing_field) {
    case 'first_name':
      updates.first_name = input
      fieldDisplayName = 'first name'
      break
    case 'last_name':
      updates.last_name = input
      fieldDisplayName = 'last name'
      break
    case 'email':
      updates.email = input
      fieldDisplayName = 'email'
      break
  }
  
  const result = await updateCustomerProfile(customerId, updates)
  
  if (!result.success) {
    const customer = await getCustomerProfile(customerId)
    const fieldName = state.editing_field === 'first_name' ? 'First Name' : 
                     state.editing_field === 'last_name' ? 'Last Name' : 'Email Address'
    const currentValue = state.editing_field === 'first_name' ? customer?.first_name :
                        state.editing_field === 'last_name' ? customer?.last_name : customer?.email
    
    const errorMessage = formatError('Failed to update profile. Please try again.')
    const fieldEdit = formatFieldEdit(fieldName, currentValue)
    return errorMessage + '\n\n' + fieldEdit
  }
  
  // Success! Show confirmation and return to profile view
  await updateConversationState(chatId, {
    ...state,
    step: 'profile_view',
    editing_field: undefined,
    previous_step: 'profile_field_edit'
  })
  
  const successMessage = formatProfileUpdated(fieldDisplayName, input)
  const updatedCustomer = await getCustomerProfile(customerId)
  
  if (updatedCustomer) {
    return successMessage + '\n\n' + formatProfileView(updatedCustomer)
  } else {
    return successMessage
  }
} 

File: backend/internal/chatbot/profile_management.go
package chatbot

import (
	"context"
	"regexp"
	"strings"
)

var emailRegex = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)

func (s *Service) handleProfileManagement(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	switch state.Step {
	case "profile_view":
		return s.handleProfileView(ctx, chatID, customerID, messageBody, state)
	case "profile_edit":
		return s.handleProfileEdit(ctx, chatID, customerID, messageBody, state)
	case "profile_field_edit":
		return s.handleProfileFieldEdit(ctx, chatID, customerID, messageBody, state)
	default:
		return s.handleProfileView(ctx, chatID, customerID, messageBody, state)
	}
}

func (s *Service) handleProfileView(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	if hasNum {
		switch selectedNumber {
		case 1:
			customer := s.getCustomerProfile(ctx, customerID)
			if customer == nil {
				return formatError("Unable to load profile. Please try again.")
			}
			newState := state
			newState.Step = "profile_edit"
			newState.PreviousStep = "profile_view"
			s.updateConversationState(ctx, chatID, newState)
			return formatProfileEdit(customer)
		case 2:
			newState := state
			newState.Step = "main_menu"
			newState.PreviousStep = "profile_view"
			s.updateConversationState(ctx, chatID, newState)
			return formatMainMenu("", 0, 0, "")
		default:
			customerForError := s.getCustomerProfile(ctx, customerID)
			if customerForError == nil {
				return formatError("Unable to load profile. Please try again.")
			}
			return formatError("Please select 1 to edit profile or 2 for main menu") + "\n\n" + formatProfileView(customerForError)
		}
	}
	customer := s.getCustomerProfile(ctx, customerID)
	if customer == nil {
		return formatError("Unable to load profile. Please try again.")
	}
	return formatProfileView(customer)
}

func (s *Service) handleProfileEdit(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	if hasNum {
		customer := s.getCustomerProfile(ctx, customerID)
		if customer == nil {
			return formatError("Unable to load profile. Please try again.")
		}
		switch selectedNumber {
		case 1:
			newState := state
			newState.Step = "profile_field_edit"
			newState.EditingField = "first_name"
			newState.PreviousStep = "profile_edit"
			s.updateConversationState(ctx, chatID, newState)
			current := ""
			if customer.FirstName != nil {
				current = *customer.FirstName
			}
			return formatFieldEdit("First Name", current)
		case 2:
			newState := state
			newState.Step = "profile_field_edit"
			newState.EditingField = "last_name"
			newState.PreviousStep = "profile_edit"
			s.updateConversationState(ctx, chatID, newState)
			current := ""
			if customer.LastName != nil {
				current = *customer.LastName
			}
			return formatFieldEdit("Last Name", current)
		case 3:
			newState := state
			newState.Step = "profile_field_edit"
			newState.EditingField = "email"
			newState.PreviousStep = "profile_edit"
			s.updateConversationState(ctx, chatID, newState)
			current := ""
			if customer.Email != nil {
				current = *customer.Email
			}
			return formatFieldEdit("Email Address", current)
		case 4:
			newState := state
			newState.Step = "profile_view"
			newState.PreviousStep = "profile_edit"
			s.updateConversationState(ctx, chatID, newState)
			return formatProfileView(customer)
		case 5:
			newState := state
			newState.Step = "main_menu"
			newState.PreviousStep = "profile_edit"
			s.updateConversationState(ctx, chatID, newState)
			return formatMainMenu("", 0, 0, "")
		default:
			return formatError("Please select a valid option (1-5)") + "\n\n" + formatProfileEdit(customer)
		}
	}
	customer := s.getCustomerProfile(ctx, customerID)
	if customer == nil {
		return formatError("Unable to load profile. Please try again.")
	}
	return formatError("Please select a valid option") + "\n\n" + formatProfileEdit(customer)
}

func (s *Service) handleProfileFieldEdit(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	input := strings.TrimSpace(messageBody)

	if strings.EqualFold(input, "cancel") {
		customer := s.getCustomerProfile(ctx, customerID)
		if customer == nil {
			return formatError("Unable to load profile. Please try again.")
		}
		newState := state
		newState.Step = "profile_edit"
		newState.EditingField = ""
		newState.PreviousStep = "profile_field_edit"
		s.updateConversationState(ctx, chatID, newState)
		return formatProfileEdit(customer)
	}

	if state.EditingField == "email" && input != "" {
		if !emailRegex.MatchString(input) {
			customer := s.getCustomerProfile(ctx, customerID)
			current := ""
			if customer != nil && customer.Email != nil {
				current = *customer.Email
			}
			return formatError("Please enter a valid email address (e.g., yourname@email.com)") + "\n\n" + formatFieldEdit("Email Address", current)
		}
	}

	if input == "" {
		customer := s.getCustomerProfile(ctx, customerID)
		fieldName := "Email Address"
		switch state.EditingField {
		case "first_name":
			fieldName = "First Name"
		case "last_name":
			fieldName = "Last Name"
		}
		var current string
		if customer != nil {
			switch state.EditingField {
			case "first_name":
				if customer.FirstName != nil {
					current = *customer.FirstName
				}
			case "last_name":
				if customer.LastName != nil {
					current = *customer.LastName
				}
			case "email":
				if customer.Email != nil {
					current = *customer.Email
				}
			}
		}
		return formatError(`Please enter a value or type "cancel" to go back`) + "\n\n" + formatFieldEdit(fieldName, current)
	}

	updates := profileUpdates{}
	fieldDisplayName := ""
	switch state.EditingField {
	case "first_name":
		updates.FirstName = &input
		fieldDisplayName = "first name"
	case "last_name":
		updates.LastName = &input
		fieldDisplayName = "last name"
	case "email":
		updates.Email = &input
		fieldDisplayName = "email"
	}

	result := s.updateCustomerProfile(ctx, customerID, updates)
	if !result.Success {
		customer := s.getCustomerProfile(ctx, customerID)
		fieldName := "Email Address"
		switch state.EditingField {
		case "first_name":
			fieldName = "First Name"
		case "last_name":
			fieldName = "Last Name"
		}
		var current string
		if customer != nil {
			switch state.EditingField {
			case "first_name":
				if customer.FirstName != nil {
					current = *customer.FirstName
				}
			case "last_name":
				if customer.LastName != nil {
					current = *customer.LastName
				}
			case "email":
				if customer.Email != nil {
					current = *customer.Email
				}
			}
		}
		return formatError("Failed to update profile. Please try again.") + "\n\n" + formatFieldEdit(fieldName, current)
	}

	newState := state
	newState.Step = "profile_view"
	newState.EditingField = ""
	newState.PreviousStep = "profile_field_edit"
	s.updateConversationState(ctx, chatID, newState)

	successMsg := formatProfileUpdated(fieldDisplayName, input)
	updatedCustomer := s.getCustomerProfile(ctx, customerID)
	if updatedCustomer != nil {
		return successMsg + "\n\n" + formatProfileView(updatedCustomer)
	}
	return successMsg
}
